packages/webamp/js/components/Vis.tsx

```diff
@@ -0,0 +1,234 @@
+import React, {
+  useMemo,
+  useCallback,
+  useState,
+  useLayoutEffect,
+  useEffect,
+} from "react";
+
+import * as Actions from "../actionCreators";
+import * as Selectors from "../selectors";
+import { FFT } from "./FFTNullsoft";
+import { useTypedSelector, useActionCreator } from "../hooks";
+// import { usePaintOscilloscopeFrame } from "./useOscilloscopeVisualizer";
+// import { usePaintBarFrame, usePaintBar } from "./useBarVisualizer";
+import { VISUALIZERS, MEDIA_STATUS } from "../constants";
+
+import {
+  Vis as IVis,
+  VisPaintHandler,
+  BarPaintHandler,
+  WavePaintHandler,
+  NoVisualizerHandler,
+} from "./VisPainter";
+
+type Props = {
+  analyser: AnalyserNode;
+};
+
+export let PIXEL_DENSITY = 1;
+
+const fft = new FFT();
+const samplesIn = 1024; // Example input size
+const samplesOut = 512; // Example output size
+export let renderWidth: number;
+export let renderHeight: number;
+export let windowShade: boolean | undefined;
+export let doubled: boolean | undefined;
+fft.init(samplesIn, samplesOut, 1, 1.0, true);
+
+let in_wavedata = new Float32Array(samplesIn); // Fill this with your input data
+export let out_spectraldata = new Float32Array(samplesOut);
+
+// Pre-render the background grid
+function preRenderBg(
+  width: number,
+  height: number,
+  bgColor: string,
+  fgColor: string,
+  windowShade: boolean
+): HTMLCanvasElement {
+  // Off-screen canvas for pre-rendering the background
+  const bgCanvas = document.createElement("canvas");
+  bgCanvas.width = width;
+  bgCanvas.height = height;
+  const distance = 2 * PIXEL_DENSITY;
+
+  const bgCanvasCtx = bgCanvas.getContext("2d");
+  if (bgCanvasCtx == null) {
+    throw new Error("Could not construct canvas context");
+  }
+  bgCanvasCtx.fillStyle = bgColor;
+  bgCanvasCtx.fillRect(0, 0, width, height);
+  if (!windowShade) {
+    bgCanvasCtx.fillStyle = fgColor;
+    for (let x = 0; x < width; x += distance) {
+      for (let y = PIXEL_DENSITY; y < height; y += distance) {
+        bgCanvasCtx.fillRect(x, y, PIXEL_DENSITY, PIXEL_DENSITY);
+      }
+    }
+  }
+  return bgCanvas;
+}
+
+export default function Vis({ analyser }: Props) {
+  useLayoutEffect(() => {
+    analyser.fftSize = 1024;
+  }, [analyser, analyser.fftSize]);
+  const colors = useTypedSelector(Selectors.getSkinColors);
+  let mode = useTypedSelector(Selectors.getVisualizerStyle);
```

Comment:
Can we make this `const` again?

Reply:
Sure, I didn't have much success in implementing the `sa` parameter (which would allow the user to set a default mode for either the analyzer, oscilloscope or nothing).